reconcilers/reconcilers.go

```diff
@@ -1163,6 +1264,89 @@ func namespaceName(obj client.Object) types.NamespacedName {
+// loaded the parent resource is used to patch it with the finalizer if not already set.
+func AddParentFinalizer(ctx context.Context, parent client.Object, finalizer string) error {
+	return ensureParentFinalizer(ctx, parent, finalizer, true)
+}
+
+// ClearParentFinalizer ensures the desired finalizer does not exist on the parent resource. The
+// client that loaded the parent resource is used to patch it with the finalizer if set.
+func ClearParentFinalizer(ctx context.Context, parent client.Object, finalizer string) error {
+	return ensureParentFinalizer(ctx, parent, finalizer, false)
+}
+
+func ensureParentFinalizer(ctx context.Context, parent client.Object, finalizer string, add bool) error {
+	config := RetrieveParentConfig(ctx)
+	if config.IsEmpty() {
+		panic(fmt.Errorf("parent config must exist on the context. Check that the context from a ParentReconciler"))
+	}
+	parentType := RetrieveParentType(ctx)
+	if parentType == nil {
+		panic(fmt.Errorf("parent type must exist on the context. Check that the context from a ParentReconciler"))
+	}
+
+	if finalizer == "" || controllerutil.ContainsFinalizer(parent, finalizer) == add {
+		// nothing to do
+		return nil
+	}
+
+	// cast the current object back to the parent so scheme-aware, typed client can operate on it
+	cast := &CastParent{
+		Type: parentType,
+		Reconciler: &SyncReconciler{
+			SyncDuringFinalization: true,
+			Sync: func(ctx context.Context, current client.Object) error {
+				log := logr.FromContextOrDiscard(ctx)
+
+				desired := current.DeepCopyObject().(client.Object)
+				if add {
+					log.Info("adding parent finalizer", "finalizer", finalizer)
+					controllerutil.AddFinalizer(desired, finalizer)
+				} else {
+					log.Info("removing parent finalizer", "finalizer", finalizer)
+					controllerutil.RemoveFinalizer(desired, finalizer)
+				}
+
+				patch := client.MergeFromWithOptions(current, client.MergeFromWithOptimisticLock{})
+				if err := config.Patch(ctx, desired, patch); err != nil {
+					log.Error(err, "unable to patch parent finalizers", "finalizer", finalizer)
+					config.Recorder.Eventf(current, corev1.EventTypeWarning, "FinalizerPatchFailed",
+						"Failed to patch finalizer %q: %s", finalizer, err)
+					return err
+				}
+				config.Recorder.Eventf(current, corev1.EventTypeNormal, "FinalizerPatched",
+					"Patched finalizer %q", finalizer)
+
+				// update current object with values from the api server after patching
+				current.SetFinalizers(desired.GetFinalizers())
+				current.SetResourceVersion(desired.GetResourceVersion())
+				current.SetGeneration(desired.GetGeneration())
+
+				return nil
+			},
+		},
+	}
+
+	_, err := cast.Reconcile(ctx, parent)
+	return err
+}
+
+// AggregateResults combines multiple results into a single result. If any result requests
+// requeue, the aggregate is requeued. The shortest non-zero requeue after is the aggregate value.
+func AggregateResults(results ...ctrl.Result) ctrl.Result {
```

Comment:
**[question, non-blocking]**: Why is `AggregateResults` exported? When would a consumer of `rr` use it?